### PR TITLE
close #1044: racecondition in `kernelShiftParticles`

### LINKED ISSUE #1044: racecondition in `kernelShiftParticles`
`cuda-memcheck --tool racecheck ` show up a race condition in the kernel `kernelShiftParticles()`.

The race condition is coming from:
https://github.com/ComputationalRadiationPhysics/picongpu/blob/master/src/libPMacc/include/particles/ParticlesBase.kernel#L184

**This bug is not critical, it is not a problem (in this case) that more than one thread wrote to the shared variable.** Nerveless it's not good to ignore this problem, therefore a fix is coming quickly.

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- remove race condition with a refactoring
- new more simple algorithm
  loops over frames:
  - step 1 - all threads: count leaving particles per direction
  - step 2 - only master threads: add frames for all leaving particles (for each direction)
  - step 3 - all threads: copy particle memory
  - step 4 - only master threads: cleanup helper variables
                                                  
This pull request close #1044.

Tests:
- [x] 3D KHI positrons/ electrons with 16 GPUs
- [x] 3D KHI positrons/ electrons with 8 GPUs

### CODE REVIEW

**Comment by ax3l** [comment]
great, thank you!
did you see a performance implication?

**Comment by psychocoderHPC** [comment]
No this pull request not increase the performance :-(

**Comment by ax3l** [comment]
Let me rephrase my question:
  I assume it decreases performance. Does it and if yes do you know how much?

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 1]
```diff
@@ -53,196 +53,150 @@ DINLINE bool getPreviousFrameAndRemoveLastFrame(typename T_ParticleBox::FrameTyp
...
+__global__ void kernelShiftParticles( ParticlesBox<FRAME, Mapping::Dim> pb, Mapping mapper )
 {
 
-    using namespace particles::operations;
-
-    /* Exchanges in 2D=8 and in 3D=26
-     */
     enum
     {
         TileSize = math::CT::volume<typename Mapping::SuperCellSize>::type::value,
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: aren't we calling this `FrameSize` "now"?

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 2]
```diff
@@ -53,196 +53,150 @@ DINLINE bool getPreviousFrameAndRemoveLastFrame(typename T_ParticleBox::FrameTyp
...
         Dim = Mapping::Dim,
+        /* Exchanges in 2D=9 and in 3D=27 */
         Exchanges = traits::NumberOfExchanges<Dim>::value
     };
 
 
-    //\todo: testen ob es schneller ist, erst zu flushen wennn Source voll ist
-    __shared__ FRAME * destFrames[Exchanges];
+    /* define memory for two times Exchanges
+     * index [0,Exchanges-1] is called first frame
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: is called [for the?] first frame

but even with the addition, I can not understand this comment. can you pls try to rephrase it?

**Comment by ax3l** [comment]
@BenjaminW3 @Heikman pls have a look :)

**Comment by psychocoderHPC** [comment]
ready to merge

**Comment by bussmann** [comment]
Nice to hear performance did not change.

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 3]
```diff
@@ -52,197 +52,147 @@ DINLINE bool getPreviousFrameAndRemoveLastFrame(typename T_ParticleBox::FrameTyp
...
+
+    if ( !mustShift || isFrameValid == false ) return;
 
-    //!\todo: find a way without so many flags
-    bool isNeighborFrame = false;
-    //init
-    if (threadIdx.x < Exchanges)
+    const DataSpace<dim> relative = superCellIdx + Mask::getRelativeDirections<dim> (linearThreadIdx + 1);
+
+    if ( linearThreadIdx < numExchanges )
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: pls add inline comments above each logical block of code such as this one

**Comment by ax3l** [comment]
thanks a bunch, very well done! :sparkles:
